File: Python/programmers/etc/level2/올바른_괄호.py
def solution(s):
    stack = []
    for i in s:
        if i == "(":
            stack.append(i)
        else:
            if stack and stack[-1] == "(":
                stack.pop()
            else:
                stack.append(")")
    return len(stack) == 0


s = "()()"
print(solution(s)) # True

s = "(())()"
print(solution(s)) # True

s = ")()("
print(solution(s)) # False

s = "(()("
print(solution(s)) # False

File: Java/programmers/etc/level2/올바른괄호.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 올바른괄호 {
    public boolean solution(String s) {
        Stack<Character> stk = new Stack<>();
        for (char i : s.toCharArray()) {
            if (i == '(') stk.push(i);
            else {
                if (!stk.isEmpty() && stk.peek() == '(') stk.pop();
                else stk.push(')');
            }
        }

        return stk.isEmpty();
    }
}
